README.md
# MicroPython I2C LCD

[![Downloads](https://pepy.tech/badge/micropython-i2c-lcd)](https://pepy.tech/project/micropython-i2c-lcd)
![Release](https://img.shields.io/github/v/release/brainelectronics/micropython-i2c-lcd?include_prereleases&color=success)
![MicroPython](https://img.shields.io/badge/micropython-Ok-green.svg)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![codecov](https://codecov.io/github/brainelectronics/micropython-i2c-lcd/branch/main/graph/badge.svg)](https://app.codecov.io/github/brainelectronics/micropython-i2c-lcd)
[![CI](https://github.com/brainelectronics/micropython-i2c-lcd/actions/workflows/release.yml/badge.svg)](https://github.com/brainelectronics/micropython-i2c-lcd/actions/workflows/release.yml)

MicroPython package to control HD44780 LCD displays 1602 and 2004 via I2C

---------------

## General

MicroPython package to control HD44780 LCD displays 1602 and 2004 via I2C

📚 The latest documentation is available at
[MicroPython I2C LCD ReadTheDocs][ref-rtd-micropython-i2c-lcd] 📚

<!-- MarkdownTOC -->

- [Installation](#installation)
	- [Install required tools](#install-required-tools)
- [Setup](#setup)
	- [Install package with upip](#install-package-with-upip)
		- [General](#general)
		- [Specific version](#specific-version)
		- [Test version](#test-version)
	- [Manually](#manually)
		- [Upload files to board](#upload-files-to-board)
- [Usage](#usage)
- [Credits](#credits)

<!-- /MarkdownTOC -->

## Installation

### Install required tools

Python3 must be installed on your system. Check the current Python version
with the following command

```bash
python --version
python3 --version
```

Depending on which command `Python 3.x.y` (with x.y as some numbers) is
returned, use that command to proceed.

```bash
python3 -m venv .venv
source .venv/bin/activate

pip install -r requirements.txt
```

## Setup

### Install package with upip

Connect the MicroPython device to a network (if possible)

```python
import network
station = network.WLAN(network.STA_IF)
station.active(True)
station.connect('SSID', 'PASSWORD')
station.isconnected()
```

#### General

Install the latest package version of this lib on the MicroPython device

```python
import mip
mip.install("github:brainelectronics/micropython-i2c-lcd")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
upip.install('micropython-i2c-lcd')
```

#### Specific version

Install a specific, fixed package version of this lib on the MicroPython device

```python
import mip
# install a verions of a specific branch
mip.install("github:brainelectronics/micropython-i2c-lcd", version="feature/initial-implementation")
# install a tag version
mip.install("github:brainelectronics/micropython-i2c-lcd", version="0.1.0")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
upip.install('micropython-i2c-lcd==0.1.0')
```

#### Test version

Install a specific release candidate version uploaded to
[Test Python Package Index](https://test.pypi.org/) on every PR on the
MicroPython device. If no specific version is set, the latest stable version
will be used.

```python
import mip
mip.install("github:brainelectronics/micropython-i2c-lcd", version="0.1.0-rc3.dev1")
```

For MicroPython versions below 1.19.1 use the `upip` package instead of `mip`

```python
import upip
# overwrite index_urls to only take artifacts from test.pypi.org
upip.index_urls = ['https://test.pypi.org/pypi']
upip.install('micropython-i2c-lcd==0.1.0rc3.dev1')
```

### Manually

#### Upload files to board

Copy the module to the MicroPython board and import them as shown below
using [Remote MicroPython shell][ref-remote-upy-shell]

Open the remote shell with the following command. Additionally use `-b 115200`
in case no CP210x is used but a CH34x.

```bash
rshell --port /dev/tty.SLAB_USBtoUART --editor nano
```

Perform the following command inside the `rshell` to copy all files and
folders to the device

```bash
mkdir /pyboard/lib
mkdir /pyboard/lib/lcd_i2c

cp lcd_i2c/* /pyboard/lib/lcd_i2c

cp examples/main.py /pyboard
cp examples/boot.py /pyboard
```

## Usage

```python
from lcd_i2c import LCD
from machine import I2C, Pin

# PCF8574 on 0x50
I2C_ADDR = 0x27     # DEC 39, HEX 0x27
NUM_ROWS = 2
NUM_COLS = 16

# define custom I2C interface, default is 'I2C(0)'
# check the docs of your device for further details and pin infos
i2c = I2C(0, scl=Pin(13), sda=Pin(12), freq=800000)
lcd = LCD(addr=I2C_ADDR, cols=NUM_COLS, rows=NUM_ROWS, i2c=i2c)

lcd.begin()
lcd.print("Hello World")
```

For further examples check the `examples` folder or the Example chapter in the
docs.

## Credits

Based on [Frank de Brabanders Arduino LiquidCrystal I2C Library][ref-arduino-lcd-i2c-library].

<!-- Links -->
[ref-rtd-micropython-i2c-lcd]: https://micropython-i2c-lcd.readthedocs.io/en/latest/
[ref-remote-upy-shell]: https://github.com/dhylands/rshell
[ref-arduino-lcd-i2c-library]: https://github.com/fdebrabander/Arduino-LiquidCrystal-I2C-library
[ref-test-pypi]: https://test.pypi.org/
[ref-pypi]: https://pypi.org/


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--
## [x.y.z] - yyyy-mm-dd
### Added
### Changed
### Removed
### Fixed
-->
<!--
RegEx for release version from file
r"^\#\# \[\d{1,}[.]\d{1,}[.]\d{1,}\] \- \d{4}\-\d{2}-\d{2}$"
-->

## Released
## [0.1.1] - 2023-06-12
### Fixed
- Usage documentation with more comments and WiFi instructions in root README
- Validate `package.json` file in test workflow

## [0.1.0] - 2023-03-09
### Added
- `const.py`, `display.py` and `typing.py` in `lcd_i2c` module
- Examples and documentation

### Changed
- Several updates on setup and config files different than the template repo

### Removed
- Not used files provided with [template repo](https://github.com/brainelectronics/micropython-i2c-lcd)

<!-- Links -->
[Unreleased]: https://github.com/brainelectronics/micropython-i2c-lcd/compare/0.1.1...main

[0.1.1]: https://github.com/brainelectronics/micropython-i2c-lcd/tree/0.1.1
[0.1.0]: https://github.com/brainelectronics/micropython-i2c-lcd/tree/0.1.0


tests/test_blink.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""Unittest for MicroPython Package template"""

import logging
from nose2.tools import params
import sys
from typing import Optional, Union
import unittest

# custom imports
from be_upy_blink import flash_led


class Pin(object):
    """Fake MicroPython Pin class"""
    IN = 1
    OUT = 2

    def __init__(self, pin: int, mode: int):
        self._pin = pin
        self._mode = mode
        self._value = 0

    def value(self, val: Optional[Union[int, bool]] = None) -> Optional[int]:
        if val is None:
            return self._value
        else:
            self._value = val


class TestBlink(unittest.TestCase):
    """This class describes a TestBlink unittest."""

    def setUp(self) -> None:
        """Run before every test method"""
        # define a format
        custom_format = "[%(asctime)s][%(levelname)-8s][%(filename)-20s @" \
                        " %(funcName)-15s:%(lineno)4s] %(message)s"

        # set basic config and level for all loggers
        logging.basicConfig(level=logging.INFO,
                            format=custom_format,
                            stream=sys.stdout)

        # create a logger for this TestSuite
        self.test_logger = logging.getLogger(__name__)

        # set the test logger level
        self.test_logger.setLevel(logging.DEBUG)

        # enable/disable the log output of the device logger for the tests
        # if enabled log data inside this test will be printed
        self.test_logger.disabled = False

    @params(
        (1),    # set pin initially HIGH
        (0),    # set pin initially LOW
    )
    def test_flash_led(self, init_state: int) -> None:
        """Test flashing of LED with fake Pin"""
        led_pin = Pin(4, Pin.OUT)
        led_pin.value(init_state)
        initial_state = led_pin.value()
        assert init_state == initial_state, "Pin not in specified init_state"

        flash_led(pin=led_pin, amount=3)

        self.assertEqual(initial_state, led_pin.value())

    def tearDown(self) -> None:
        """Run after every test method"""
        pass


if __name__ == '__main__':
    unittest.main()


tests/test_lcd_i2c.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""Unittest for MicroPython I2C LCD"""

import logging
from unittest.mock import Mock, patch
from nose2.tools import params
import sys
import unittest


class Pin(object):
    """Fake MicroPython Pin class"""
    def __init__(self, pin: int, mode: int = -1):
        self._pin = pin
        self._mode = mode
        self._value = 0


class I2C(object):
    """Fake MicroPython I2C class"""
    def __init__(self, id: int, *, scl: Pin, sda: Pin, freq: int = 400000):
        self._id = id
        self._scl = scl
        self._sda = sda
        self._freq = freq

    def writeto(addr: int, buf: bytearray, stop: bool = True) -> int:
        return 1


# custom imports
sys.modules['machine.I2C'] = I2C
to_be_mocked = [
    'machine',
    'time.sleep_ms', 'time.sleep_us',
]
for module in to_be_mocked:
    sys.modules[module] = Mock()

from lcd_i2c import LCD             # noqa: E402
from lcd_i2c import const as Const  # noqa: E402


class TestLCD(unittest.TestCase):
    """This class describes a TestLCD unittest."""

    def setUp(self) -> None:
        """Run before every test method"""
        # define a format
        custom_format = "[%(asctime)s][%(levelname)-8s][%(filename)-20s @" \
                        " %(funcName)-15s:%(lineno)4s] %(message)s"

        # set basic config and level for all loggers
        logging.basicConfig(level=logging.INFO,
                            format=custom_format,
                            stream=sys.stdout)

        # create a logger for this TestSuite
        self.test_logger = logging.getLogger(__name__)

        # set the test logger level
        self.test_logger.setLevel(logging.DEBUG)

        # enable/disable the log output of the device logger for the tests
        # if enabled log data inside this test will be printed
        self.test_logger.disabled = False

        self.i2c = I2C(1, scl=Pin(3), sda=Pin(2), freq=800_000)
        self._calls_counter = 0
        self._tracked_call_data: list = []

    def _tracked_call(self, *args, **kwargs) -> None:
        """Track function calls and the used arguments"""
        self._tracked_call_data.append({'args': args, 'kwargs': kwargs})

    @params(
        (0x27),
        (39),
    )
    def test_addr(self, addr: int) -> None:
        """Test address property"""
        lcd = LCD(addr=addr, cols=16, rows=2, i2c=self.i2c)

        self.assertEqual(lcd.addr, addr)

    @params(
        (16),
        (20),
    )
    def test_cols(self, cols: int) -> None:
        """Test columns property"""
        lcd = LCD(addr=0x27, cols=cols, rows=2, i2c=self.i2c)

        self.assertEqual(lcd.cols, cols)

    @params(
        (1),
        (2),
        (4),
    )
    def test_rows(self, rows: int) -> None:
        """Test rows property"""
        lcd = LCD(addr=0x27, cols=16, rows=rows, i2c=self.i2c)

        self.assertEqual(lcd.rows, rows)

    def test_charsize(self) -> None:
        """Test charsize property"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        self.assertEqual(lcd.charsize, 0)

        lcd = LCD(addr=0x27, cols=16, rows=2, charsize=1, i2c=self.i2c)
        self.assertEqual(lcd.charsize, 1)

    def test_backlightval(self) -> None:
        """Test backlightval property"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)

        # active by default
        self.assertEqual(lcd.backlightval, Const.LCD_BACKLIGHT)
        self.assertTrue(lcd.get_backlight())

        lcd.no_backlight()
        self.assertEqual(lcd.backlightval, Const.LCD_NOBACKLIGHT)
        self.assertFalse(lcd.get_backlight())

        lcd.backlight()
        self.assertEqual(lcd.backlightval, Const.LCD_BACKLIGHT)
        self.assertTrue(lcd.get_backlight())

        lcd.set_backlight(new_val=False)
        self.assertEqual(lcd.backlightval, Const.LCD_NOBACKLIGHT)
        self.assertFalse(lcd.get_backlight())

        lcd.set_backlight(new_val=True)
        self.assertEqual(lcd.backlightval, Const.LCD_BACKLIGHT)
        self.assertTrue(lcd.get_backlight())

    def test_cursor_position(self) -> None:
        """Test cursor position property"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        self.assertEqual(lcd.cursor_position, (0, 0))

        # test valid cursor position
        lcd.cursor_position = (10, 1)
        self.assertEqual(lcd.cursor_position, (10, 1))

        # test invald cursor position (not enough rows)
        lcd.cursor_position = (10, 2)
        self.assertEqual(lcd.cursor_position, (10, 1))

    def test_begin(self) -> None:
        """Test LCD begin"""
        # default I2C interface
        lcd = LCD(addr=0x27, cols=16, rows=2)
        lcd.begin()

        # self.assertEqual(lcd._i2c._id, 0)
        self.assertEqual(lcd._display_function, 0x8)
        self.assertEqual(lcd._display_control, 0x4)
        self.assertEqual(lcd._display_mode, 2)
        self.assertEqual(lcd.cursor_position, (0, 0))

        # double row display
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()

        # self.assertEqual(lcd._i2c._id, self.i2c._id)
        self.assertEqual(lcd._display_function, 0x8)
        self.assertEqual(lcd._display_control, 0x4)
        self.assertEqual(lcd._display_mode, 2)
        self.assertEqual(lcd.cursor_position, (0, 0))

        # single row display
        lcd = LCD(addr=0x27, cols=16, rows=1, i2c=self.i2c)
        lcd.begin()

        # self.assertEqual(lcd._i2c._id, self.i2c._id)
        self.assertEqual(lcd._display_function, 0x0)
        self.assertEqual(lcd._display_control, 0x4)
        self.assertEqual(lcd._display_mode, 2)
        self.assertEqual(lcd.cursor_position, (0, 0))

        # single row display with different char size
        lcd = LCD(addr=0x27, cols=16, rows=1, charsize=0x1, i2c=self.i2c)
        lcd.begin()

        # self.assertEqual(lcd._i2c._id, self.i2c._id)
        self.assertEqual(lcd._display_function, 0x4)
        self.assertEqual(lcd._display_control, 0x4)
        self.assertEqual(lcd._display_mode, 2)
        self.assertEqual(lcd.cursor_position, (0, 0))

    def test_clear(self) -> None:
        """Test clear display"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.set_cursor(col=3, row=1)

        lcd.cursor_position = (3, 1)
        lcd.clear()
        lcd.cursor_position = (0, 0)

    def test_home(self) -> None:
        """Test home display"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.set_cursor(col=3, row=1)

        lcd.cursor_position = (3, 1)
        lcd.home()
        lcd.cursor_position = (0, 0)

    def test_display(self) -> None:
        """Test display on/off functions"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()
        self.assertEqual(lcd._display_control, 0x4)

        lcd.no_display()
        self.assertEqual(lcd._display_control, 0x0)

        lcd.display()
        self.assertEqual(lcd._display_control, 0x4)

    def test_blink(self) -> None:
        """Test cursor blink on/off functions"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()
        self.assertEqual(lcd._display_control, 0x4)

        lcd.blink()
        self.assertEqual(lcd._display_control, 0x5)

        lcd.no_blink()
        self.assertEqual(lcd._display_control, 0x4)

        lcd.blink_on()
        self.assertEqual(lcd._display_control, 0x5)

        lcd.blink_off()
        self.assertEqual(lcd._display_control, 0x4)

    def test_cursor(self) -> None:
        """Test cursor on/off functions"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()
        self.assertEqual(lcd._display_control, 0x4)

        lcd.cursor()
        self.assertEqual(lcd._display_control, 0x6)

        lcd.no_cursor()
        self.assertEqual(lcd._display_control, 0x4)

        lcd.cursor_on()
        self.assertEqual(lcd._display_control, 0x6)

        lcd.cursor_off()
        self.assertEqual(lcd._display_control, 0x4)

        lcd.cursor()
        lcd.blink()
        self.assertEqual(lcd._display_control, 0x7)

        lcd.no_blink()
        self.assertEqual(lcd._display_control, 0x6)

        lcd.no_cursor()
        self.assertEqual(lcd._display_control, 0x4)

    def test_scroll_display_left(self) -> None:
        """Test scrolling test to the left"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()

        with patch('lcd_i2c.LCD._command', wraps=self._tracked_call):
            lcd.scroll_display_left()

        self.assertEqual(len(self._tracked_call_data), 1)
        self.assertEqual(self._tracked_call_data[0]['kwargs']['value'], 0x18)

    def test_scroll_display_right(self) -> None:
        """Test scrolling test to the right"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()

        with patch('lcd_i2c.LCD._command', wraps=self._tracked_call):
            lcd.scroll_display_right()

        self.assertEqual(len(self._tracked_call_data), 1)
        self.assertEqual(self._tracked_call_data[0]['kwargs']['value'], 0x1C)

    def test_text_flow(self) -> None:
        """Test setting text flow left to right"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()
        self.assertEqual(lcd._display_mode, 0x2)

        lcd.right_to_left()
        self.assertEqual(lcd._display_mode, 0x0)

        lcd.left_to_right()
        self.assertEqual(lcd._display_mode, 0x2)

    def test_autoscroll(self) -> None:
        """Test autoscroll function"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()
        self.assertEqual(lcd._display_mode, 0x2)

        lcd.autoscroll()
        self.assertEqual(lcd._display_mode, 0x3)

        lcd.no_autoscroll()
        self.assertEqual(lcd._display_mode, 0x2)

    def test_create_char(self) -> None:
        """Test creating custom char"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()

        charmap = [
            0x0,    # 00000
            0x1,    # 00001
            0x3,    # 00011
            0x7,    # 00111
            0xF,    # 01111
            0x1F,   # 11111
            0x4,    # 00100
            0x11,   # 10001
        ]

        with patch('lcd_i2c.LCD._command', wraps=self._tracked_call):
            lcd.create_char(location=0, charmap=charmap)

        self.assertEqual(len(self._tracked_call_data) - 1, len(charmap))

        # command value to set CGRAM address at the given location
        self.assertEqual(self._tracked_call_data.pop(0)['kwargs']['value'], 64)

        # check char map send command content
        for idx, val in enumerate(charmap):
            self.assertEqual(self._tracked_call_data[idx]['kwargs']['value'],
                             val)

    def test_print(self) -> None:
        """Test print on LCD"""
        lcd = LCD(addr=0x27, cols=16, rows=2, i2c=self.i2c)
        lcd.begin()
        self.assertEqual(lcd.cursor_position, (0, 0))

        text = "Hello"

        with patch('lcd_i2c.LCD._command', wraps=self._tracked_call):
            lcd.print(text)

        # called 1x to much
        self.assertEqual(len(self._tracked_call_data) - 1, len(text))
        self.assertEqual(lcd.cursor_position, (0 + len(text), 0))

        for idx, val in enumerate(text):
            self.assertEqual(self._tracked_call_data[idx]['kwargs']['value'],
                             ord(val))

    def tearDown(self) -> None:
        """Run after every test method"""
        pass


if __name__ == '__main__':
    unittest.main()


tests/test_absolute_truth.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
"""Unittest for testing the absolute truth"""

import logging
from nose2.tools import params
import sys
from typing import Any
import unittest


class TestAbsoluteTruth(unittest.TestCase):
    def setUp(self) -> None:
        """Run before every test method"""
        # define a format
        custom_format = "[%(asctime)s][%(levelname)-8s][%(filename)-20s @" \
                        " %(funcName)-15s:%(lineno)4s] %(message)s"

        # set basic config and level for all loggers
        logging.basicConfig(level=logging.INFO,
                            format=custom_format,
                            stream=sys.stdout)

        # create a logger for this TestSuite
        self.test_logger = logging.getLogger(__name__)

        # set the test logger level
        self.test_logger.setLevel(logging.DEBUG)

        # enable/disable the log output of the device logger for the tests
        # if enabled log data inside this test will be printed
        self.test_logger.disabled = False

    def test_absolute_truth(self) -> None:
        """Test the unittest itself"""
        x = 0
        y = 1
        z = 2
        none_thing = None
        some_dict = dict()
        some_list = [x, y, 40, "asdf", z]

        self.assertTrue(True)
        self.assertFalse(False)

        self.assertEqual(y, 1)
        assert y == 1
        self.assertNotEqual(x, y)
        assert x != y

        self.assertIsNone(none_thing)
        self.assertIsNotNone(some_dict)

        self.assertIn(y, some_list)
        self.assertNotIn(12, some_list)

        # self.assertRaises(exc, fun, args, *kwds)

        self.assertIsInstance(some_dict, dict)
        self.assertNotIsInstance(some_dict, list)

        self.assertGreater(a=y, b=x)
        self.assertGreaterEqual(a=y, b=x)
        self.assertLess(a=x, b=y)

        self.test_logger.debug("Sample debug message")

    @params(
        (123.45, True),
        (1, False)
    )
    def test_with_params(self, parameter: Any, expectation: bool) -> None:
        """
        Test something using parameters

        :param      parameter:    The parameter value
        :type       parameter:    Any
        :param      expectation:  The expectation
        :type       expectation:  bool
        """
        if expectation:
            self.assertIsInstance(parameter, float)
        else:
            self.assertNotIsInstance(parameter, float)

    def tearDown(self) -> None:
        """Run after every test method"""
        pass


if __name__ == '__main__':
    unittest.main()


lcd_i2c/const.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

from micropython import const

# commands
#: Clear display command
LCD_CLEARDISPLAY = const(0x01)
#: Return to home position command
LCD_RETURNHOME = const(0x02)
#: Set entry mode command
LCD_ENTRYMODESET = const(0x04)
#: Control display command
LCD_DISPLAYCONTROL = const(0x08)
#: Shift cursor command
LCD_CURSORSHIFT = const(0x10)
#: Set function command
LCD_FUNCTIONSET = const(0x20)
#: Set CGRAM address command
LCD_SETCGRAMADDR = const(0x40)
#: Set DDRAM address command
LCD_SETDDRAMADDR = const(0x80)

# flags for display entry mode
#: Set display entry mode as right command
LCD_ENTRYRIGHT = const(0x00)
#: Set display entry mode as left command
LCD_ENTRYLEFT = const(0x02)
#: Set display entry mode as shift increment command
LCD_ENTRYSHIFTINCREMENT = const(0x01)
#: Set display entry mode as shift decrement command
LCD_ENTRYSHIFTDECREMENT = const(0x00)

# flags for display on/off control
#: Turn display on command
LCD_DISPLAYON = const(0x04)
#: Turn display off command
LCD_DISPLAYOFF = const(0x00)
#: Turn cursor on command
LCD_CURSORON = const(0x02)
#: Turn cursor off command
LCD_CURSOROFF = const(0x00)
#: Set curor blink command
LCD_BLINKON = const(0x01)
#: Set curor no blink command
LCD_BLINKOFF = const(0x00)

# flags for display/cursor shift
#: Display move command
LCD_DISPLAYMOVE = const(0x08)
#: Move cursor command
LCD_CURSORMOVE = const(0x00)
#: Move display shift right command
LCD_MOVERIGHT = const(0x04)
#: Move display shift left command
LCD_MOVELEFT = const(0x00)

# flags for function set
#: 8 bit mode command
LCD_8BITMODE = const(0x10)
#: 4 bit mode command
LCD_4BITMODE = const(0x00)
#: 2 line command
LCD_2LINE = const(0x08)
#: 1 line command
LCD_1LINE = const(0x00)
#: 5x10 dots display command
LCD_5x10DOTS = const(0x04)
#: 5x8 dots display command
LCD_5x8DOTS = const(0x00)

# flags for backlight control
#: Activate backlight command
LCD_BACKLIGHT = const(0x08)
#: Deactivate backlight command
LCD_NOBACKLIGHT = const(0x00)

# other
#: Enable bit
EN = const(0b00000100)
#: Read/Write bit
RW = const(0b00000010)
#: Register select bit
RS = const(0b00000001)


be_upy_blink/blink.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""
Very simple blink helper module

Toggle given pin on and off with some delay
"""

from time import sleep


def flash_led(pin, amount, on_time=0.5, off_time=0.5):
    """
    Flash onboard LED at given pin

    :param      pin:       The pin connected to the LED
    :type       pin:       int
    :param      amount:    The amount the LED flashes
    :type       amount:    int
    :param      on_time:   On time of the LED
    :type       on_time:   Union[float, int]
    :param      off_time:  Off time of the LED
    :type       off_time:  Union[float, int]
    """
    initial_state = pin.value()

    for x in range(1, amount + 1):
        pin.value(1)
        sleep(on_time)
        pin.value(0)
        sleep(off_time)

    pin.value(initial_state)


examples/main.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""I2C LCD showcase"""

from lcd_i2c import LCD
from machine import I2C, Pin
from time import sleep

# PCF8574 on 0x27
I2C_ADDR = 0x27     # DEC 39, HEX 0x27
I2C_NUM_ROWS = 2
I2C_NUM_COLS = 16
FREQ = 800000   # Try lowering this value in case of "Errno 5"


def print_and_wait(text: str, sleep_time: int = 2) -> None:
    """
    Print to console and wait some time.

    :param      text:        The text to print to console
    :type       text:        str
    :param      sleep_time:  The sleep time in seconds
    :type       sleep_time:  int
    """
    print(text)
    sleep(sleep_time)


# define custom I2C interface, default is 'I2C(0)'
# check the docs of your device for further details and pin infos
# this are the pins for the Raspberry Pi Pico adapter board
i2c = I2C(0, scl=Pin(13), sda=Pin(12), freq=FREQ)
lcd = LCD(addr=I2C_ADDR, cols=I2C_NUM_COLS, rows=I2C_NUM_ROWS, i2c=i2c)

# get LCD infos/properties
print("LCD is on I2C address {}".format(lcd.addr))
print("LCD has {} columns and {} rows".format(lcd.cols, lcd.rows))
print("LCD is used with a charsize of {}".format(lcd.charsize))
print("Cursor position is {}".format(lcd.cursor_position))

# start LCD, not automatically called during init to be Arduino compatible
lcd.begin()

# print text on sceen at first row, starting on first column
lcd.print("Hello World")
print_and_wait("Show 'Hello World' on LCD")

# turn LCD off
lcd.no_backlight()
print_and_wait("Turn LCD backlight off")

# get current backlight value
print("Backlight value: {}".format(lcd.get_backlight()))

# turn LCD on
lcd.backlight()
print_and_wait("Turn LCD backlight on")

# get current backlight value
print("Backlight value: {}".format(lcd.get_backlight()))

# clear LCD display content
lcd.clear()
print_and_wait("Clear display content")

# turn cursor on (show)
lcd.cursor()
print_and_wait("Turn cursor on (show)")

# blink cursor
lcd.blink()
print_and_wait("Blink cursor")

# return cursor to home position
lcd.home()
print_and_wait("Return cursor to home position")

# stop blinking cursor
lcd.no_blink()
print_and_wait("Stop blinking cursor")

# turn cursor off (hide)
lcd.no_cursor()
print_and_wait("Turn cursor off (hide)")

# print_and_wait text on sceen
lcd.print("Hello again")
print_and_wait("Show 'Hello again' on LCD")

# turn display off
lcd.no_display()
print_and_wait("Turn LCD off")

# turn display on
lcd.display()
print_and_wait("Turn LCD on")

# scroll display to the left
for _ in "Hello again":
    lcd.scroll_display_left()
    sleep(0.5)
print_and_wait("Scroll display to the left")

# scroll display to the right
for _ in "Hello again":
    lcd.scroll_display_right()
    sleep(0.5)
print_and_wait("Scroll display to the right")

# set text flow right to left
lcd.clear()
lcd.set_cursor(col=12, row=0)
lcd.right_to_left()
lcd.print("Right to left")
print_and_wait("Set text flow right to left")

# set text flow left to right
lcd.clear()
lcd.set_cursor(col=0, row=0)
lcd.left_to_right()
lcd.print("Left to right")
print_and_wait("Set text flow left to right")

# activate autoscroll
lcd.autoscroll()
print_and_wait("Activate autoscroll")

# disable autoscroll
lcd.no_autoscroll()
print_and_wait("Disable autoscroll")

# set cursor to second line, seventh column
lcd.clear()
lcd.cursor()
# lcd.cursor_position = (7, 1)
lcd.set_cursor(col=7, row=1)
print_and_wait("Set cursor to row 1, column 7")
lcd.no_cursor()

# set custom char number 0 as :-)
# custom char can be set for location 0 ... 7
lcd.create_char(
    location=0,
    charmap=[0x00, 0x00, 0x11, 0x04, 0x04, 0x11, 0x0E, 0x00]
    # this is the binary matrix, feel it, see it
    # 00000
    # 00000
    # 10001
    # 00100
    # 00100
    # 10001
    # 01110
    # 00000
)
print_and_wait("Create custom char ':-)'")

# show custom char stored at location 0
lcd.print(chr(0))
lcd.print(chr(0))
print_and_wait("Show custom char")
